wen-restart/src/wen_restart.rs

```diff
@@ -196,7 +232,100 @@ pub(crate) fn aggregate_restart_last_voted_fork_slots(
             sleep(Duration::from_millis(time_left));
         }
     }
-    Ok(())
+    Ok(last_voted_fork_slots_aggregate.get_final_result())
+}
+
+pub(crate) fn find_heaviest_fork(
+    aggregate_final_result: LastVotedForkSlotsFinalResult,
+    bank_forks: Arc<RwLock<BankForks>>,
+    exit: Arc<AtomicBool>,
+) -> Result<(Slot, Hash)> {
+    // Under normal cases without duplicate blocks, we should have a single chain of
+    // blocks from the root to the newest block with > (67% - 5% -  (100% - X%)) of stake.
+    // 1. If X% of the validators joined the restart (X > 80%), then there can't be two
+    //    kids B and C of the same parent A with more than (67% - 5% -  (100% - X%)) of
+    //    the stake each. Assume B and C exists, their combined stake is more than
+    //    2 * (67% - 5% - (100% - X%)) = 134% - 10% - 2 * (100% - X%) = 2*X% - 76% > X%.
+    //    Since we can only have X% of the stake, this is a contradiction.
+    // 2. If there are no duplicate blocks and a child got Y% of the stake, then its
+    //    parent should at least have Y% of the stake.
+    // So, we can start from the latest block with > (67% - 5% -  (100% - X%)) = X% - 38%
```

Comment:
Hmm, I don't think such a child necessarily exists?

Say a validator V contributed to voting on the last optimistically confirmed slot X.  They could have then voted on some other  last slot that Y>X such that Y does not have 42% of the stake

Reply:
Moved the whole discussion to SIMD.